Source/MockDefaultValueProvider.cs

```diff
@@ -59,23 +60,40 @@ public MockDefaultValueProvider(Mock owner)
 
 		public override object ProvideDefault(MethodInfo member)
 		{
-			var value = base.ProvideDefault(member);
+			Type taskResultType = null;
 
-			Mock mock = null;
+			var returnType = member.ReturnType;
+			if (returnType.GetTypeInfo().IsGenericType)
+			{
+				var returnTypeGenericDef = returnType.GetGenericTypeDefinition();
+				if (returnTypeGenericDef == typeof(Task<>))
+				{
+					taskResultType = returnType.GetGenericArguments()[0];
+				}
+			}
+
+			object value = null;
+			if (taskResultType == null)
+			{
+				value = base.ProvideDefault(member);
+			}
+
+			MockWithOptionalWrapper mock = default(MockWithOptionalWrapper);
 			if (value == null && member.ReturnType.IsMockeable())
 			{
 				mock = owner.InnerMocks.GetOrAdd(member, info =>
 				{
 					// Create a new mock to be placed to InnerMocks dictionary if it's missing there
-					var mockType = typeof(Mock<>).MakeGenericType(info.ReturnType);
+					var mockType = typeof(Mock<>).MakeGenericType(taskResultType ?? info.ReturnType);
 					Mock newMock = (Mock)Activator.CreateInstance(mockType, owner.Behavior);
 					newMock.DefaultValue = owner.DefaultValue;
 					newMock.CallBase = owner.CallBase;
-					return newMock;
+					var wrapper = taskResultType != null ? CreateCompletedTask(taskResultType, newMock.Object) : null;
+					return new MockWithOptionalWrapper(newMock, wrapper);
 				});
 			}
 
-			return mock != null ? mock.Object : value;
+			return mock.Mock != null ? mock.Wrapper ?? mock.Mock.Object : value;
```

Comment:
Doesn't read very clearly. Probably the same as `mock.Wrapper ?? mock.Mock?.Object ?? value`, but it would be better to rewrite the whole method so that no convoluted expression is required at its end.